File: solutions/java/sgf-parsing/3/src/main/java/SgfParsing.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class SgfParsing {

    public SgfNode parse(String input) throws SgfParsingException {
        if (input.isBlank() || input.charAt(0) != '(' || input.charAt(input.length() - 1) != ')') {
            throw new SgfParsingException("tree missing");
        }
        ParserCursor cursor = new ParserCursor();
        return parseTree(input, cursor);
    }

    private SgfNode parseTree(String input, ParserCursor cursor) throws SgfParsingException {
        if (input.charAt(cursor.index) != '(') {
            throw new SgfParsingException("expected '(' at position " + cursor.index);
        }
        cursor.index++;

        List<SgfNode> sequence = new ArrayList<>();
        while (cursor.index < input.length() && input.charAt(cursor.index) != ')') {
            char c = input.charAt(cursor.index);
            if (c == ';') {
                cursor.index++;
                sequence.add(parseNode(input, cursor));
            } else if (c == '(') {
                SgfNode variation = parseTree(input, cursor);
                if (!sequence.isEmpty()) {
                    sequence.get(sequence.size() - 1).appendChild(variation);
                }
            } else {
                cursor.index++; 
            }
        }

        if (cursor.index >= input.length() || input.charAt(cursor.index) != ')') {
            throw new SgfParsingException("expected ')' at position " + cursor.index);
        }
        cursor.index++;

        if (sequence.isEmpty()) {
            throw new SgfParsingException("tree with no nodes");
        }

        SgfNode root = sequence.get(0);
        SgfNode current = root;
        for (int i = 1; i < sequence.size(); i++) {
            current.appendChild(sequence.get(i));
            current = sequence.get(i);
        }

        return root;
    }

    private SgfNode parseNode(String input, ParserCursor cursor) throws SgfParsingException {
        StringBuilder nodeBuilder = new StringBuilder();
        boolean inBracket = false;
        boolean escaping = false;

        while (cursor.index < input.length()) {
            char c = input.charAt(cursor.index);
            if (!inBracket && (c == ';' || c == '(' || c == ')')) {
                break;
            }

            if (escaping) {
                nodeBuilder.append(c);
                escaping = false;
            } else if (c == '\\') {
                nodeBuilder.append(c);
                escaping = true;
            } else {
                nodeBuilder.append(c);
                if (c == '[') inBracket = true;
                if (c == ']') inBracket = false;
            }

            cursor.index++;
        }

        List<String> properties = getPropertiesNode(nodeBuilder.toString());
        return new SgfNode(parsePropertiesMap(properties));
    }

    private List<String> getPropertiesNode(String node) {
        List<String> result = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inBracket = false;
        boolean escaping = false;

        for (int i = 0; i < node.length(); i++) {
            char c = node.charAt(i);

            if (escaping) {
                current.append(c);
                escaping = false;
                continue;
            }

            if (c == '\\') {
                current.append(c);
                escaping = true;
                continue;
            }

            if (c == '[') {
                inBracket = true;
                current.append(c);
                continue;
            }

            if (c == ']') {
                inBracket = false;
                current.append(c);
                continue;
            }

            if (!inBracket && Character.isUpperCase(c)) {
                if (!current.isEmpty()) {
                    result.add(current.toString());
                    current.setLength(0);
                }
            }

            current.append(c);
        }

        if (!current.isEmpty()) {
            result.add(current.toString());
        }

        return result;
    }

    public Map<String, List<String>> parsePropertiesMap(List<String> nodes) throws SgfParsingException {
        Map<String, List<String>> properties = new LinkedHashMap<>();
        for (String node : nodes) {
            if (node.isEmpty()) continue;
            int i = 0;
            while (i < node.length() && node.charAt(i) != '[') {
                i++;
            }
            String key = node.substring(0, i);
            if (!key.equals(key.toUpperCase())) {
                throw new SgfParsingException("property must be in uppercase");
            }
            if (key.length() == node.length()) {
                throw new SgfParsingException("properties without delimiter");
            }

            List<String> values = new ArrayList<>();

            while (i < node.length()) {
                if (node.charAt(i) == '[') {
                    i++;
                    StringBuilder value = new StringBuilder();
                    boolean escaping = false;

                    while (i < node.length()) {
                        char c = node.charAt(i);
                        if (escaping) {
                            if (c == '\n' || c == '\r') {
                            } else {
                                value.append(c);
                            }
                            escaping = false;
                        } else if (c == '\\') {
                            escaping = true;
                        } else if (c == ']') {
                            break;
                        } else {
                            value.append(c);
                        }
                        i++;
                    }
                    values.add(value.toString().replace("\t"," "));
                }
                i++;
            }
            properties.put(key, values);
        }
        return properties;
    }

    private static class ParserCursor {
        int index = 0;
    }
}

File: solutions/python/sgf-parsing/1/sgf_parsing.py
class SgfTree:
    def __init__(self, properties=None, children=None):
        self.properties = properties or {}
        self.children = children or []

    def __eq__(self, other):
        if not isinstance(other, SgfTree):
            return False
        for key, value in self.properties.items():
            if key not in other.properties:
                return False
            if other.properties[key] != value:
                return False
        for key in other.properties.keys():
            if key not in self.properties:
                return False
        if len(self.children) != len(other.children):
            return False
        for child, other_child in zip(self.children, other.children):
            if child != other_child:
                return False
        return True

    def __ne__(self, other):
        return not self == other


def parse(input_string):
    if not input_string:
        raise ValueError("tree missing")
    if input_string == "()":
        raise ValueError("tree with no nodes")
    if input_string[0] != "(" or input_string[-1] != ")":
        raise ValueError("tree missing")

    index = [1]

    def parse_tree():
        if index[0] >= len(input_string) or input_string[index[0]] != ';':
            raise ValueError("tree missing")

        return parse_nodes()

    def parse_nodes():
        if input_string[index[0]] != ';':
            raise ValueError("tree missing")
        index[0] += 1

        properties = {}
        while index[0] < len(input_string):
            char = input_string[index[0]]

            if char.isupper():
                prop_key = ""

                while index[0] < len(input_string) and input_string[index[0]].isalpha():
                    prop_key += input_string[index[0]]
                    index[0] += 1

                if not prop_key.isupper():
                    raise ValueError("property must be in uppercase")

                if index[0] >= len(input_string) or input_string[index[0]] != '[':
                    raise ValueError("properties without delimiter")

                values = []
                while index[0] < len(input_string) and input_string[index[0]] == '[':
                    index[0] += 1
                    val = ""
                    escaped = False
                    while index[0] < len(input_string):
                        c = input_string[index[0]]
                        if escaped:
                            if c == '\n':
                                pass
                            elif c in ['\t', ' ']:
                                val += ' '
                            else:
                                val += c
                            escaped = False
                        elif c == '\\':
                            escaped = True
                        elif c == ']':
                            index[0] += 1
                            break
                        else:
                            val += ' ' if c == '\t' else c
                        index[0] += 1
                    values.append(val)
                properties[prop_key] = values

            elif char == ';' or char == '(' or char == ')':
                break
            elif char.islower():
                raise ValueError("property must be in uppercase")
            else:
                index[0] += 1

        children = []
        if index[0] < len(input_string) and input_string[index[0]] == ';':
            children.append(parse_nodes())
        while index[0] < len(input_string) and input_string[index[0]] == '(':
            index[0] += 1
            children.append(parse_nodes())
            if input_string[index[0]] != ')':
                raise ValueError("tree missing")
            index[0] += 1

        return SgfTree(properties, children)

    result = parse_tree()

    if index[0] < len(input_string) - 1:
        index[0] += 1

    return result